### PR TITLE
Switch DYAD to use RMA instead of 2 sided communication for better performance

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
image

- [x] init and finalize rkey
- [x] switch 2 sided to rma
- [x] use rma to send size and data
- [x] Add Compile time option to switch between 2-sided and RMA. (RMA is default)

### CODE REVIEW

**Comment by JaeseungYeom** [comment]
I am familiar with all but "2-sided" part. Can you explain that? Is the RMA scheme still the same as when we looked at it last time or different?

**Comment by hariharan-devarajan** [comment]
2 sided communication means we have an active send and recv for communication. 

For rdma yes, it's same as we discussed. The consumers sends a streaming rpc to broker 
The brokers writes the data into consumers buffer.

**Comment by hariharan-devarajan** [comment]
@JaeseungYeom The PR is ready. 
1. One issue to add for further improvement is to cache the remote key, and its buffer. 
2. Also, I don't think Flux_RPC CI works so I am disabling it for now. 
But functionally, this PR is complete.

**Comment by hariharan-devarajan** [comment]
@JaeseungYeom This PR is ready for review

File: src/dyad/dtl/ucx_dtl.c [hunk 1]
```diff
@@ -214,10 +219,10 @@ ucx_free_buf_done:;
     return rc;
 }
 
-static inline ucs_status_ptr_t ucx_send_no_wait (const dyad_ctx_t* ctx, void* buf, size_t buflen)
+static inline ucs_status_ptr_t ucx_send_no_wait (const dyad_ctx_t* ctx, bool is_warmup, void* buf, size_t buflen)
```
Comment (JaeseungYeom) [inline src/dyad/dtl/ucx_dtl.c:222]: Need a comment to explain the behavior with this argument.

File: src/dyad/dtl/ucx_dtl.c [hunk 2]
```diff
@@ -4,17 +4,19 @@
...
+#include <dyad/dtl/ucx_dtl.h>
 #include <dyad/utils/base64/base64.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <assert.h>
-
+#include <unistd.h>
+#define _OPEN_SYS_ITOA_EXT
```
Comment (JaeseungYeom) [inline src/dyad/dtl/ucx_dtl.c:17]: What is this? Need some comment

File: .github/workflows/compile_test.yaml [hunk 3]
```diff
@@ -12,7 +12,7 @@ jobs:
       fail-fast: false
       matrix:
         flux: [ 0.52.0, 0.49.0]
-        mode: ["FLUX_RPC", "UCX"]
+        mode: ["UCX"]
```
Comment (JaeseungYeom) [inline .github/workflows/compile_test.yaml:15]: Create an issue regarding FLUX_RPC failure. Do you see it fail even when you run manually? It worked for me recently.

File: src/dyad/dtl/ucx_dtl.c [hunk 4]
```diff
@@ -509,10 +555,8 @@ dyad_rc_t dyad_dtl_ucx_init (const dyad_ctx_t* ctx,
     //   * Auto initialization of request objects
     //   * Worker sleep, wakeup, poll, etc. features
     ucx_params.field_mask =
-        UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE | UCP_PARAM_FIELD_REQUEST_INIT;
-    ucx_params.features = UCP_FEATURE_TAG |
-                          // UCP_FEATURE_RMA |
-                          UCP_FEATURE_WAKEUP;
+        UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE;
+    ucx_params.features = UCP_FEATURE_RMA | UCP_FEATURE_AMO32;;
```
Comment (JaeseungYeom) [inline src/dyad/dtl/ucx_dtl.c:559]: Does this work for non RMA case?

**Review by JaeseungYeom — COMMENTED** [review]
Need some clarifications.

File: .github/workflows/compile_test.yaml (hunk 3 above)
Comment (hariharan-devarajan) [reply .github/workflows/compile_test.yaml:15]: Yes I got the same error as the CI, Maybe Me and the CI are missing something.

File: src/dyad/modules/dyad.c [hunk 5]
```diff
@@ -186,8 +186,10 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
     file_size = get_file_size (fd);
     DYAD_LOG_DEBUG (mod_ctx->ctx, "file %s has size %u", fullpath, file_size);
     if (file_size > 0) {
-        rc = mod_ctx->ctx->dtl_handle->get_buffer (mod_ctx->ctx, file_size, &inbuf);
-        inlen = read (fd, inbuf, file_size);
+        size_t fs = file_size;
+        rc = mod_ctx->ctx->dtl_handle->get_buffer (mod_ctx->ctx, file_size, (void**)&inbuf);
+        memcpy (inbuf, &fs, sizeof(fs));
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/dyad/modules/dyad.c:197]: Added comments.

File: src/dyad/dtl/ucx_dtl.c (hunk 2 above)
Comment (hariharan-devarajan) [reply src/dyad/dtl/ucx_dtl.c]: Oh i was trying a API for serialization. Cleaned it up.

File: src/dyad/dtl/ucx_dtl.c (hunk 4 above)
Comment (hariharan-devarajan) [reply src/dyad/dtl/ucx_dtl.c]: Added the TAG back. Somehow it got deleted. Oops

File: src/dyad/dtl/ucx_dtl.c (hunk 1 above)
Comment (hariharan-devarajan) [reply src/dyad/dtl/ucx_dtl.c:220]: Added

**Review by JaeseungYeom — APPROVED** [review]
Looks good to me.

**Pull request merged**